app/src/org/commcare/tasks/ProcessAndSendTask.java

```diff
@@ -174,60 +177,65 @@ private boolean checkFormRecordStatus(FormRecord[] records)
 
             //If the form is complete, but unprocessed, process it.
             if (FormRecord.STATUS_COMPLETE.equals(record.getStatus())) {
+                SQLiteDatabase db =
+                        CommCareApplication.instance().getUserDbHandle();
+                db.beginTransaction();
                 try {
                     records[i] = processor.process(record);
+                    db.setTransactionSuccessful();
                 } catch (InvalidStructureException | XmlPullParserException |
                         UnfullfilledRequirementsException e) {
-                    handleExceptionFromFormProcessing(record, e);
+                    db.endTransaction();

```

Comment:
The reason all of these calls to `endTransaction()` need to be separated out (instead of being called just once in a `finally` block) is that the transaction needs to be ended _before_ the catch blocks execute their responses to the exceptions, which in some cases make db modifications that we do NOT want to be rolled back

Reply:
Makes sense. Another option would be to use a `finally` block that uses `inTransaction` to check whether a previous `catch` has closed the transaction already and only closes if not. Probably doesn't make this much cleaner

Reply:
@wpride That pattern is actually quite dangerous because it wont' be able to identify a nested transaction (so you could still end up firing two "ends" for one "begin"